### PR TITLE
Fixing completion water cut checking for WTEST and some related refactoring

### PR DESCRIPTION
**Pull request by GitPaean**
Mostly addressing two issues was discovered when developing https://github.com/OPM/opm-simulators/pull/1882 . 

One is related to performance related to completions.  The other one is related to the calculation of completion water cut based on the rates of each connections.  And an function is introduced to reduce the duplicated code when calculating the water cut. 

We begin to use connection instead of perforation with this PR. Connection will be the more proper name numerically. 

fixing here refer to the comment https://github.com/OPM/opm-simulators/pull/1882/files#r292866208

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Since there were not review done, I pushed more changes, which still staying within the scope of the PR title. 

Some efforts were done to make this part easier to extend to handle different limits without duplicating a lot of code.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
In term of functionality, the only change is to fix the logic to handle the rate ratio for completions.   Others are refactoring for the later development.  Fortunately, this PR does not fail the related test. 

Later development will begin focusing on functionality development and extension. I believe it is good to get this PR reviewed and merged at its current stage.

**Comment by GitPaean** [comment]
The current refactoring is quite locally to the function `checkMaxWaterCutLimit` without touching other design.  With later development, more refactoring of the current structure for WellTesting be join step by step.

**Comment by GitPaean** [comment]
jenkins build this please

**Review by totto82 — APPROVED** [review]
The refactoring seems reasonable. Thanks. Merge when green

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/simulators/wells/WellInterface_impl.hpp [hunk 1]
```diff
@@ -718,6 +705,78 @@ namespace Opm
 
 
 
+    template<typename TypeTag>
+    bool
+    WellInterface<TypeTag>::
+    checkMaxRatioLimitWell(const WellState& well_state,
+                           const double max_ratio_limit,
+                           double (*ratioFunc)(const std::vector<double>&, const PhaseUsage&)) const
```
Comment (atgeirr) [inline opm/simulators/wells/WellInterface_impl.hpp]: No need to require function pointers here, it is hard to read and prevents inlining. Rather, make this a template member taking a RatioFunctor template argument. The calling code should not require changes.

**Review by atgeirr — CHANGES_REQUESTED** [review]
Looks good except for one noted issue (template instead of using function pointers).

Do we have regression tests that verify the behaviour of the refactored code? Do you get bit-identical behaviour?

**Comment by GitPaean** [comment]
> Do we have regression tests that verify the behaviour of the refactored code? 

Yes. 3D_WECON test.

> Do you get bit-identical behaviour?

Yes. Unfortunately.  The logic for the completion water cut changes, while for this case, it does not change the behavoir.

File: opm/simulators/wells/WellInterface_impl.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/simulators/wells/WellInterface_impl.hpp]: Please check it is what you want. I do not use it very often.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
I think all the comments are addressed already. Please let me know if you have more comments. 

Otherwise, I think this is ready to go in. More PRs are coming soon.

File: opm/simulators/wells/WellInterface_impl.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/simulators/wells/WellInterface_impl.hpp]: Updated with more communications.

**Comment by GitPaean** [comment]
jenkins build this please

**Review by atgeirr — APPROVED** [review]

**Comment by atgeirr** [comment]
Thanks, merging.

**Comment by bska** [comment]
This PR **appears** to break [Norne](https://ci.opm-project.org/job/opm-simulators/17/console):
```
Report step 10/247 at day 168/3312, date = 23-Apr-1998

Time step 0, stepsize 8 days, at day 168/176, date = 23-Apr-1998
flow: /var/lib/jenkins/post-builder/workspace/opm-simulators/opm/simulators/wells/WellInterface_impl.hpp:153: void Opm::WellInterface<TypeTag>::initCompletions() [with TypeTag = Ewoms::Properties::TTag::EclFlowProblem]: Assertion `num_conns == number_of_perforations_' failed.
jenkins/run-norne.sh: line 10: 13757 Aborted                 $WORKSPACE/$configuration/build-opm-simulators/bin/$SIM --output-dir=$SIM NORNE_ATW2013.DATA
Build step 'Execute shell' marked build as failure
Archiving artifacts
Recording test results
```
Is that just an artefact of the CI system or do anyone else see similar symptoms?

**Comment by bska** [comment]
> This PR **appears** to break [Norne](https://ci.opm-project.org/job/opm-simulators/17/console):
> 
> [...]
> 
> Is that just an artefact of the CI system or do anyone else see similar symptoms?

Update: I see the same on my local system when running a debug build of Flow.

**Comment by GitPaean** [comment]
Sorry for that. 

It looks like the assumption `Assertion `num_conns == number_of_perforations_' ` does not apply for Norne.  I am investigating it.

**Comment by GitPaean** [comment]
` well B-4H num_conns 20 number_of_perforations_ 3`

while the COMPDAT for well `B-4H`, 

```
 388 COMPDAT ^M
 389 -- WELL        I    J    K1  K2            Sat.        CF       DIAM        KH SKIN ND        DIR   Ro ^M
 390      'B-4H'   10   32    1    1      'OPEN'  1*    118.457      0.216  12025.229  2*         'Z'     24.860 /^M
 391      'B-4H'   10   32    2    2      'OPEN'  1*     88.058      0.216   8892.405  2*         'Z'     24.161 /^M
 392      'B-4H'   10   32    3    3      'OPEN'  1*      6.493      0.216    651.729  2*         'Z'     23.389 /^M
 393      'B-4H'   10   32    5    5      'OPEN'  1*     37.039      0.216   3688.499  2*         'Z'     22.416 /^M
 394      'B-4H'   10   32    6    6      'OPEN'  1*     37.182      0.216   3691.211  2*         'Z'     22.046 /^M
 395      'B-4H'   10   32    7    7      'OPEN'  1*     27.449      0.216   2717.094  2*         'Z'     21.709 /^M
 396      'B-4H'   10   32    8    8      'OPEN'  1*     27.553      0.216   2719.652  2*         'Z'     21.384 /^M
 397      'B-4H'   10   32    9    9      'OPEN'  1*     55.934      0.216   5504.213  2*         'Z'     21.045 /^M
 398      'B-4H'   10   32   10   10      'OPEN'  1*    343.826      0.216  33643.313  2*         'Z'     20.427 /^M
 399      'B-4H'    9   32   13   13      'OPEN'  1*    247.048      0.216  23303.623  2*         'Z'     16.915 /^M
 400      'B-4H'    9   32   14   14      'OPEN'  1*    200.684      0.216  18964.795  2*         'Z'     17.072 /^M
 401      'B-4H'    9   32   15   15      'OPEN'  1*     60.756      0.216   5750.007  2*         'Z'     17.201 /^M
 402      'B-4H'    9   32   16   16      'OPEN'  1*    131.035      0.216  12417.905  2*         'Z'     17.318 /^M
 403      'B-4H'    9   32   17   17      'OPEN'  1*      1.215      0.216    115.284  2*         'Z'     17.399 /^M
 404      'B-4H'    9   32   18   18      'OPEN'  1*    122.230      0.216  11601.419  2*         'Z'     17.455 /^M
 405      'B-4H'    9   32   19   19      'OPEN'  1*     50.908      0.216   4838.953  2*         'Z'     17.585 /^M
 406      'B-4H'    9   32   20   20      'OPEN'  1*     84.765      0.216   8082.179  2*         'Z'     17.865 /^M
 407      'B-4H'    9   31   20   20      'OPEN'  1*     65.987      0.216   6333.601  2*         'Z'     18.482 /^M
 408      'B-4H'    9   31   21   21      'OPEN'  1*     40.421      0.216   3901.629  2*         'Z'     19.028 /^M
 409      'B-4H'    9   31   22   22      'OPEN'  1*    440.439      0.216  42866.883  2*         'Z'     19.863 /^M
 410 /^
```

**Comment by GitPaean** [comment]
It is possible a lot of connections are not open. I will check more.

**Comment by bska** [comment]
> It is possible a lot of connections are not open. I will check more.

Or that the connection is in a deactivated cell.

**Comment by GitPaean** [comment]
> Or that the connection is in a deactivated cell.

nope. those ones are removed from the EclipseState already.  There are some CLOSE connections.  Basically the algorithm to address COMPLUMP for WECON before or after this PR does not consider the situation that CLOSE connections does not enter WellsManager.  It is assumed that all the connections are OPEN. 

Working on a patch now.

**Comment by bska** [comment]
> > Or that the connection is in a deactivated cell.
> 
> nope. those ones are removed from the EclipseState already.

If you say so.  I was only going by the `COMPDAT` data you cited for `B-4H` which defines 20 connections.

**Comment by GitPaean** [comment]
>  I was only going by the COMPDAT data you cited for B-4H which defines 20 connections.

Sorry, I should have posted more.  There are some WELOPEN and COMPDAT for well `B-4H` later.

**Comment by GitPaean** [comment]
Some information outputting for the well `B-4H`
```
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 1
 connection open ? 1
 connection open ? 1
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
 connection open ? 0
```

**Pull request merged**